exchanges/request/request.go

```diff
@@ -237,3 +261,13 @@ func (r *Requester) SetProxy(p *url.URL) error {
 	}
 	return nil
 }
+
+func (r *Requester) drainBody(body io.ReadCloser) {
+	defer body.Close()
+	if _, err := io.Copy(ioutil.Discard, io.LimitReader(body, drainBodyLimit)); err != nil {
```

Comment:
If all we are doing is draining, probably just read 0 or 1 byte instead of 100,000? Just say the body is bigger than 100k does this not drain? 
If we fail to drain as well can we send/re-use another request, might want to return that as an error and then stop a new potential request that will fail?

Reply:
io.LimitReader will read to N bytes then return EOF so it will still drain but since we are not doing anything still most likely not worth draining up to 100,000 bytes

Reply:
100,000 bytes was chosen arbitrarily, probably needs to be adjusted based on what the typical response sizes are for cases that may need to be retried.

> Just say the body is bigger than 100k does this not drain?

Correct: a body larger than 100k will not fully drain, so a 'keep alive' connection won't be able to be reused.

In regards to the draining, my understanding is that unless the bytes are _actually_ read from the HTTP response body, the HTTP connection cannot be reused. Having a smaller limit reader means we read less of the HTTP body, and thus are more likely to not read the entire response (thus not be able to reuse the connection). This is true when we haven't read the entire response body, even if we close it.

```
// [...] It is the caller's responsibility to
// close Body. The default HTTP client's Transport may not
// reuse HTTP/1.x "keep-alive" TCP connections if the Body is
// not read to completion and closed.
```

The limit reader returning EOF doesn't mean that the entire HTTP response body has been read, as the limit reader just wraps the original reader, and returns EOF if at/beyond the limit: https://github.com/golang/go/blob/master/src/io/io.go#L450

Reply:
> 100,000 bytes was chosen arbitrarily, probably needs to be adjusted based on what the typical response sizes are for cases that may need to be retried.

Thanks for clarifying, to be honest the error body going above and beyond 100k bytes to me looks like a low probability/issue anyway, but sending a request if it fails we can handle so if you can return an error here that would be better. 

@thrasher- for more input.

Reply:
> [...] but sending a request if it fails we can handle so if you can return an error here that would be better.

I'm not sure that returning an error here is the desired behaviour, but I assume that isn't what is intended. If I'm understanding what the intention is, you would prefer _not_ draining the body at all and just closing it, having a new HTTP connection created for the next request?

Would a smaller limit be ok as an alternative? If errors mostly aren't larger than 5k (or maybe even smaller?), maybe that is a better limit to apply and have _most_ cases drain successfully, while anything larger than that will not drain fully, and force a new connection to be established in those cases.

Reply:
I think the limit being 100k is fine. It will ensure that most of response bodies will be drained and a connection is allowed to be reused. The reader will stop anyway if an EOF is hit for anything smaller. I was going to request you change it to just being `io.Copy(ioutil.Discard, body)` but there's a chance of a massive body that causes delays. 100k bytes isn't very much

Also, I don't think an error needs to be returned. If draining the body fails because the response body had an unexpected EOF or that the body is too large, there's not much we can action other than increasing the limit, log that it occurred or do nothing. It just means that we'll have to make a new connection